### PR TITLE
Allow complete_reindex lambda to handle missing version

### PR DESCRIPTION
**Pull request by kenoir**
### What is this PR trying to achieve?

Allow complete_reindex lambda to handle missing version!

### Who is this change for?

📴 

### Have the following been considered/are they needed?

- [ ] Deployed new versions
- [ ] Run `terraform apply`.

### CODE REVIEW

**Comment by kenoir** [comment]
Supersedes: https://github.com/wellcometrust/platform/pull/1552

File: reindexer/complete_reindex/src/complete_reindex.py [hunk 1]
```diff
@@ -2,31 +2,80 @@
...
-def main(event, _ctxt=None, dynamodb_client=None):
-    print(f'event = {event!r}')
-    dynamodb_client = dynamodb_client or boto3.client('dynamodb')
+def _get_item_version(dynamo_item):
+    dynamo_item_version = 1
+
+    if 'version' in dynamo_item:
+        dynamo_item_version = dynamo_item['version']
+
+    return dynamo_item_version
```
Comment (alexwlchan) [inline reindexer/complete_reindex/src/complete_reindex.py]: Easier would be to use the `get` method on Python dicts: `dict.get(key[, default]`

i.e.

```python
dynamo_item.get('version', 1)
```

File: reindexer/complete_reindex/src/complete_reindex.py [hunk 2]
```diff
@@ -2,31 +2,80 @@
...
+    dynamodb_response = table.get_item(Key={'shardId': shard_id})
 
-    table_name = os.environ['TABLE_NAME']
+    dynamo_item = dynamodb_response['Item']
 
+    print(f'Retrieved {dynamo_item}')
+
+    dynamo_current_version = dynamo_item['currentVersion']
+
+    if (dynamo_current_version >= completed_reindex_version):
```
Comment (alexwlchan) [inline reindexer/complete_reindex/src/complete_reindex.py]: nit: drop the parens

File: reindexer/complete_reindex/src/complete_reindex.py [hunk 3]
```diff
@@ -2,31 +2,80 @@
...
-                }
-            )
-        except ClientError as err:
-            if err.response['Error']['Code'] == 'ConditionalCheckFailedException':
-                print(f'{row["shardId"]} already has a newer currentVersion')
-            else:
-                raise
+        job = _process_reindex_tracker_update_job(table, record.message)
+
+    if job is not None:
```
Comment (alexwlchan) [inline reindexer/complete_reindex/src/complete_reindex.py]: Should this be in the body of the loop? – you construct a job for every update, but only process the last one. Possibly needs extra tests?

File: reindexer/complete_reindex/src/complete_reindex.py [hunk 4]
```diff
@@ -2,31 +2,80 @@
...
+
+    return dynamo_item_version
+
+
+def _update_versioned_item(table, dynamo_item, desired_item):
+    desired_dynamo_conditional_update_version = (
+            _get_item_version(dynamo_item) + 1
+    )
+
+    new_dynamo_item = {**dynamo_item, **desired_item}
```
Comment (alexwlchan) [inline reindexer/complete_reindex/src/complete_reindex.py]: This construction is quite unusual – could we just pass in the item we want to PUT?

**Pull request merged**